src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs

```diff
@@ -78,6 +80,27 @@ private static bool IsValidRootConfigType(ITypeSymbol? type)
                 }
 
                 return true;
+
+                static bool IsAccessibleFromGenBinders(ITypeSymbol type)
+                {
+                    if (type is IArrayTypeSymbol array)
+                    {
+                        return IsAccessibleFromGenBinders(array.ElementType);
+                    }
+
+                    if (type is INamedTypeSymbol namedType && namedType.GetAllTypeArgumentsInScope() is List<ITypeSymbol> typeArgsInScope)
+                    {
+                        foreach (ITypeSymbol genericArg in typeArgsInScope)
+                        {
+                            if (!IsAccessibleFromGenBinders(genericArg))
+                            {
+                                return false;
+                            }
+                        }
+                    }
+
+                    return type.DeclaredAccessibility is Accessibility.Public or Accessibility.Internal or Accessibility.Friend;
```

Comment:
What about `internal` members that are in referenced assemblies?

Reply:
We shouldn't be seeing those in this routine. This method is for root binding types only. `InternalsVisibleTo` usage is interesting but the current logic handles the scenario.

For transitive types - the generator only inspects public properties. The compiler would have complained if a type had a public member whose declared type is internal. So we don't have to guard against that explicitly

Reply:
My point is we should try to use IsSymbolAccessibleWithin rather than attempt to reimplement accessibility resolution.

Reply:
Yeah I fully intended to use it; will do if possible per discussion here https://github.com/dotnet/runtime/pull/91657#discussion_r1318574410.